apollo-federation/src/sources/connect/validation/snapshots/validation_tests@invalid_namespace_in_body_selection.graphql.snap

```diff
@@ -0,0 +1,14 @@
+---
+source: apollo-federation/src/sources/connect/validation/mod.rs
+expression: "format!(\"{:#?}\", errors)"
+input_file: apollo-federation/src/sources/connect/validation/test_data/invalid_namespace_in_body_selection.graphql
+---
+[
+    Message {
+        code: InvalidJsonSelection,
+        message: "`@connect(http: {body:})` on `Mutation.createUser` contains an invalid variable namespace `$status`, must be one of $args, $config, $context, $this",
+        locations: [
+            14:9..14:16,
```

Comment:
Oof, looks like this test failed on windows since the range is 14:8..14:15 there. Will need to investigate this.

Reply:
The problem here was that JSON `JSONSelection::parse` takes `&str` as the input parameter. I had mistakenly passed in the result of calling `as_str()` on `Node<Value>`, instead of a the result of calling `as_str()` on a `GraphQLString` created from the `Node<Value>`. The difference is that the `Node<Value>` is from the parsed GraphQL AST, which has the [GraphQL block string value rules](https://spec.graphql.org/October2021/#BlockStringValue()) applied - so it has `\r` removed, and has the common indent whitespace removed from the start of the lines. The `GraphQLString`, on the other hand, points directly to the raw string data from the schema file, which has the `\r` and all leading whitespace in tact.

When doing validation, or anything else that will use the location offsets returned by the parser, you must parse the raw string data from a `GraphQLString`, rather than the GraphQL AST data. If you do the latter, any location offsets returned by the parser will be completely invalid with respect to the original schema file.

I've fixed the mistake, but there may be something better we can do here. Rust has the ability to make bugs like this impossible. But we aren't taking advantage of the language type system, and are instead just working with bare `&str`, which has no type information to enforce what that reference is actually pointing to. So the compiler can't help prevent this kind of mistake. This isn't particular to JSON selection - the URL template and header parsers also take `&str`.